harness/tests/experiment/pytorch/test_pytorch_trial.py

```diff
@@ -636,3 +666,77 @@ def test_checkpoint_loading(ckpt: str, istrial: bool):
         assert isinstance(trial, pytorch.PyTorchTrial), type(trial)
     else:
         assert isinstance(trial, torch.nn.Module), type(trial)
+
+
+def make_amp_workloads(
+    trial_class,
+    assert_output_float16=False,
+) -> workload.Stream:
+    trainer = utils.TrainAndValidate()
+    yield from trainer.send(steps=20, validation_freq=1, scheduling_unit=1)
+    training_metrics, _ = trainer.result()
+
+    for metrics, next_metrics in zip(training_metrics[:-1], training_metrics[1:]):
+        scale = next_metrics["scale_before"]
+        if "scale" in metrics:
+            assert metrics["scale"] == scale
+        else:
+            metrics["scale"] = scale
+        assert "scale" in metrics
+    loss_prev = None
+    GROWTH_INTERVAL = trial_class._growth_interval
+    growth_countdown = GROWTH_INTERVAL
+    # Only attempt assertions up to and including the penultimate batch, because
+    #  we may not have the updated scale from the final batch.
+    for idx, metrics in enumerate(training_metrics[:-1]):
```

Comment:
does it make sense/is there a good way to break up the checks inside this for loop into their own functions? like separate methods for checking dtype, scale, scale_loss, etc. it's fine as is, but would make it easier to read if there was some way to organize the logic.

Reply:
I think we were right on the edge of whether or not it'd be worth making more methods in order to reduce the body of another. After combining the loops as you suggested (73a6008) and reversing the assertion logic (8004a18), I think this is less of a concern (still iffy, though).